File: src/main/java/uk/co/eelpieconsulting/feedlistener/controllers/SubscriptionsController.java
package uk.co.eelpieconsulting.feedlistener.controllers;

import com.google.common.base.Strings;
import com.mongodb.MongoException;
import org.apache.log4j.Logger;
import org.json.JSONException;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.servlet.ModelAndView;
import org.springframework.web.servlet.view.RedirectView;
import uk.co.eelpieconsulting.common.geo.model.LatLong;
import uk.co.eelpieconsulting.common.http.HttpFetchException;
import uk.co.eelpieconsulting.common.views.ViewFactory;
import uk.co.eelpieconsulting.feedlistener.CredentialsRequiredException;
import uk.co.eelpieconsulting.feedlistener.UnknownSubscriptionException;
import uk.co.eelpieconsulting.feedlistener.UrlBuilder;
import uk.co.eelpieconsulting.feedlistener.annotations.Timed;
import uk.co.eelpieconsulting.feedlistener.daos.ChannelsDAO;
import uk.co.eelpieconsulting.feedlistener.daos.SubscriptionsDAO;
import uk.co.eelpieconsulting.feedlistener.daos.UsersDAO;
import uk.co.eelpieconsulting.feedlistener.exceptions.UnknownUserException;
import uk.co.eelpieconsulting.feedlistener.instagram.InstagramSubscriptionManager;
import uk.co.eelpieconsulting.feedlistener.model.*;
import uk.co.eelpieconsulting.feedlistener.rss.RssPoller;
import uk.co.eelpieconsulting.feedlistener.rss.RssSubscriptionManager;
import uk.co.eelpieconsulting.feedlistener.twitter.TwitterListener;
import uk.co.eelpieconsulting.feedlistener.twitter.TwitterSubscriptionManager;

import javax.servlet.http.HttpServletResponse;
import java.io.UnsupportedEncodingException;
import java.net.UnknownHostException;

@Controller
public class SubscriptionsController {

    private final static Logger log = Logger.getLogger(SubscriptionsController.class);

    private UsersDAO usersDAO;
    private SubscriptionsDAO subscriptionsDAO;
    private RssPoller rssPoller;
    private TwitterListener twitterListener;
    private InstagramSubscriptionManager instagramSubscriptionManager;
    private UrlBuilder urlBuilder;
    private TwitterSubscriptionManager twitterSubscriptionManager;
    private RssSubscriptionManager rssSubscriptionManager;
    private ViewFactory viewFactory;
    private FeedItemPopulator feedItemPopulator;

    public SubscriptionsController() {
    }

    @Autowired
    public SubscriptionsController(UsersDAO usersDAO, SubscriptionsDAO subscriptionsDAO, RssPoller rssPoller, TwitterListener twitterListener,
                                   InstagramSubscriptionManager instagramSubscriptionManager, UrlBuilder urlBuilder,
                                   TwitterSubscriptionManager twitterSubscriptionManager,
                                   RssSubscriptionManager rssSubscriptionManager,
                                   ViewFactory viewFactory, FeedItemPopulator feedItemPopulator) {
        this.usersDAO = usersDAO;
        this.subscriptionsDAO = subscriptionsDAO;
        this.rssPoller = rssPoller;
        this.twitterListener = twitterListener;
        this.instagramSubscriptionManager = instagramSubscriptionManager;
        this.urlBuilder = urlBuilder;
        this.twitterSubscriptionManager = twitterSubscriptionManager;
        this.rssSubscriptionManager = rssSubscriptionManager;
        this.viewFactory = viewFactory;
        this.feedItemPopulator = feedItemPopulator;
    }

    @Timed(timingNotes = "")
    @RequestMapping(value = "/{username}/subscriptions/{id}/items", method = RequestMethod.GET)
    public ModelAndView subscriptionItems(@PathVariable String username, @PathVariable String id,
                                          @RequestParam(required = false) Integer page,
                                          @RequestParam(required = false) String format,
                                          HttpServletResponse response) throws MongoException, UnknownSubscriptionException, UnknownUserException {
        usersDAO.getByUsername(username);

        Subscription subscription = subscriptionsDAO.getById(username, id);

        ModelAndView mv = new ModelAndView(viewFactory.getJsonView());
        if (!Strings.isNullOrEmpty(format) && format.equals("rss")) {
            String title = !Strings.isNullOrEmpty(subscription.getName()) ? subscription.getName() : subscription.getId();
            mv = new ModelAndView(viewFactory.getRssView(title, urlBuilder.getSubscriptionUrl(subscription), ""));
        }

        long totalCount = feedItemPopulator.populateFeedItems(subscription, page, mv, "data");
        response.addHeader("X-Total-Count", Long.toString(totalCount));
        return mv;
    }

    @Timed(timingNotes = "")
    @RequestMapping(value = "/{username}/subscriptions/{id}/reload", method = RequestMethod.GET)
    public ModelAndView reload(@PathVariable String username, @PathVariable String id) throws MongoException, UnknownSubscriptionException, UnknownUserException {
        usersDAO.getByUsername(username);

        RssSubscription subscription = (RssSubscription) subscriptionsDAO.getById(username, id);

        log.info("Requesting reload of subscription: " + subscription.getName() + " / " + subscription.getUrl());
        rssPoller.run(subscription);

        ModelAndView mv = new ModelAndView(viewFactory.getJsonView());
        mv.addObject("data", "ok");
        return mv;
    }

    @Timed(timingNotes = "")
    @RequestMapping(value = "/subscriptions/{id}", method = RequestMethod.GET)
    public ModelAndView subscriptionJson(@PathVariable String username, @PathVariable String id,
                                         @RequestParam(required = false) Integer page) throws MongoException, UnknownSubscriptionException, UnknownUserException {
        usersDAO.getByUsername(username);

        Subscription subscription = subscriptionsDAO.getById(username, id);

        final ModelAndView mv = new ModelAndView(viewFactory.getJsonView());
        feedItemPopulator.populateFeedItems(subscription, page, mv, "data");    // TODO this looks incorrect
        return mv;
    }

    @Timed(timingNotes = "")
    @RequestMapping(value = "/{username}/subscriptions/{id}/delete")    // TODO should be a HTTP DELETE
    public ModelAndView deleteSubscription(@PathVariable String username, @PathVariable String id) throws MongoException, HttpFetchException, UnknownSubscriptionException, UnknownUserException {
        usersDAO.getByUsername(username);

        Subscription subscription = subscriptionsDAO.getById(username, id);
        if (subscription == null) {
            // TODO 404
            return null;
        }

        // feedItemDAO.deleteSubscriptionFeedItems(subscription); TODO
        subscriptionsDAO.delete(subscription);

        if (subscription.getId().startsWith("twitter")) {
            twitterListener.connect();
        }
        if (subscription.getId().startsWith("instagram")) {
            instagramSubscriptionManager.requestUnsubscribeFrom(((InstagramSubscription) subscription).getSubscriptionId());
        }

        return new ModelAndView(viewFactory.getJsonView()).addObject("data", "ok");
    }

    @Timed(timingNotes = "")
    @RequestMapping(value = "/{username}/subscriptions", method = RequestMethod.GET)
    public ModelAndView subscriptions(@PathVariable String username, @RequestParam(required = false) String url) throws UnknownUserException {
        usersDAO.getByUsername(username);

        final ModelAndView mv = new ModelAndView(viewFactory.getJsonView());
        mv.addObject("data", subscriptionsDAO.getSubscriptions(SubscriptionsDAO.LATEST_ITEM_DATE_DESCENDING, url));
        return mv;
    }

    @Timed(timingNotes = "")
    @RequestMapping(value = "/{username}/subscriptions/feeds", method = RequestMethod.POST)
    public ModelAndView addFeedSubscription(@PathVariable String username, @RequestParam String url, @RequestParam String channel) throws UnknownUserException {
        usersDAO.getByUsername(username);

        RssSubscription subscription = rssSubscriptionManager.requestFeedSubscription(url, channel, username);
        subscriptionsDAO.add(subscription);
        log.info("Added subscription: " + subscription);

        rssPoller.run(subscription);

        return new ModelAndView(new RedirectView(urlBuilder.getSubscriptionUrl(subscription)));
    }

    @Timed(timingNotes = "")
    @RequestMapping(value = "/{username}/subscriptions/twitter/tags", method = RequestMethod.POST)
    public ModelAndView addTwitterTagSubscription(@PathVariable String username, @RequestParam String tag, @RequestParam String channel) throws CredentialsRequiredException, UnknownUserException {
        log.info("Twitter tag: " + tag);
        twitterSubscriptionManager.requestTagSubscription(tag, channel, username);

        return new ModelAndView(new RedirectView(urlBuilder.getBaseUrl()));
    }

    @Timed(timingNotes = "")
    @RequestMapping(value = "/{username}/subscriptions/instagram/tags", method = RequestMethod.POST)
    public ModelAndView addInstagramTagSubscription(@PathVariable String username,
                                                    @RequestParam String tag, @RequestParam String channel) throws UnsupportedEncodingException, HttpFetchException, JSONException, UnknownUserException {
        log.info("Instagram tag");
        InstagramSubscription subscription = instagramSubscriptionManager.requestInstagramTagSubscription(tag, channel, username);
        subscriptionsDAO.add(subscription);

        return new ModelAndView(new RedirectView(urlBuilder.getBaseUrl()));
    }

    @Timed(timingNotes = "")
    @RequestMapping(value = "/{username}/subscriptions/instagram/geography", method = RequestMethod.POST)
    public ModelAndView addInstagramTagSubscription(@PathVariable String username,
                                                    @RequestParam double latitude,
                                                    @RequestParam double longitude,
                                                    @RequestParam int radius,
                                                    @RequestParam String channel) throws UnsupportedEncodingException, HttpFetchException, JSONException, UnknownUserException {
        LatLong latLong = new LatLong(latitude, longitude);

        InstagramGeographySubscription instagramGeographySubscription = instagramSubscriptionManager.requestInstagramGeographySubscription(latLong, radius, channel, username);
        log.info("Saving subscription: " + instagramGeographySubscription);
        subscriptionsDAO.add(instagramGeographySubscription);

        return new ModelAndView(new RedirectView(urlBuilder.getBaseUrl()));
    }

}

File: src/main/kotlin/uk/co/eelpieconsulting/feedlistener/controllers/SubscriptionsController.kt
package uk.co.eelpieconsulting.feedlistener.controllers

import com.google.common.base.Strings
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.apache.logging.log4j.LogManager
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.HttpStatus
import org.springframework.stereotype.Controller
import org.springframework.web.bind.annotation.*
import org.springframework.web.server.ResponseStatusException
import org.springframework.web.servlet.ModelAndView
import uk.co.eelpieconsulting.feedlistener.UrlBuilder
import uk.co.eelpieconsulting.feedlistener.controllers.ui.WithSignedInUser
import uk.co.eelpieconsulting.feedlistener.daos.FeedItemDAO
import uk.co.eelpieconsulting.feedlistener.daos.SubscriptionsDAO
import uk.co.eelpieconsulting.feedlistener.model.RssSubscription
import uk.co.eelpieconsulting.feedlistener.model.User
import uk.co.eelpieconsulting.feedlistener.rss.RssPoller
import uk.co.eelpieconsulting.feedlistener.views.ViewFactory
import java.net.MalformedURLException
import java.net.URL

@Controller
class SubscriptionsController @Autowired constructor(
    private val subscriptionsDAO: SubscriptionsDAO,
    private val feedItemPopulator: FeedItemPopulator,
    private val feedItemDAO: FeedItemDAO,
    private val viewFactory: ViewFactory,
    private val urlBuilder: UrlBuilder,
    private val rssPoller: RssPoller,
    private val conditionalLoads: ConditionalLoads,
    currentUserService: CurrentUserService,
    request: HttpServletRequest
) : WithSignedInUser(currentUserService, request) {

    private val log = LogManager.getLogger(ChannelsController::class.java)

    private val xTotalCount = "X-Total-Count"

    @CrossOrigin
    @GetMapping("/subscriptions/{id}/items")
    fun subscriptionItems(
        @PathVariable id: String,
        @RequestParam(required = false) page: Int?,
        @RequestParam(required = false) pageSize: Int?,
        @RequestParam(required = false) format: String?,
        response: HttpServletResponse
    ): ModelAndView {

        fun renderSubscriptionItems(user: User): ModelAndView {
            return conditionalLoads.withSubscriptionForUser(id, user) { subscription ->
                val mv = if (!Strings.isNullOrEmpty(format) && format == "rss") {
                    val title = if (!Strings.isNullOrEmpty(subscription.name)) subscription.name else subscription.id
                    ModelAndView(viewFactory.rssView(title, urlBuilder.getSubscriptionUrl(subscription), ""))
                } else {
                    ModelAndView(viewFactory.jsonView())
                }
                val feedItemsResult = feedItemDAO.getSubscriptionFeedItems(subscription, page, pageSize)
                feedItemPopulator.populateFeedItems(feedItemsResult, mv, "data")
                response.addHeader(xTotalCount, feedItemsResult.totalCount.toString())
                mv
            }
        }

        return forCurrentUser(::renderSubscriptionItems)
    }

    @GetMapping("/subscriptions/{id}/read") // TODO Make a POST
    fun reload(@PathVariable id: String): ModelAndView {
        return forCurrentUser { user ->
            conditionalLoads.withSubscriptionForUser(id, user) { subscription ->
                if (subscription is RssSubscription) {
                    rssPoller.requestRead(subscription)
                    ModelAndView(viewFactory.jsonView()).addObject("data", "ok")
                } else {
                    throw ResponseStatusException(HttpStatus.BAD_REQUEST, "Subscription is not an RSS feed")
                }
            }
        }
    }

    @CrossOrigin
    @GetMapping("/subscriptions/{id}")
    fun subscriptionJson(
        @PathVariable id: String,
        @RequestParam(required = false) page: Int?
    ): ModelAndView {
        fun renderSubscription(user: User): ModelAndView {
            return conditionalLoads.withSubscriptionForUser(id, user) { subscription ->
                ModelAndView(viewFactory.jsonView()).addObject("data", subscription)
            }
        }
        return forCurrentUser(::renderSubscription)
    }

    @PostMapping("/subscriptions")
    fun createSubscription(@RequestBody create: SubscriptionCreateRequest): ModelAndView {
        fun createSubscription(user: User): ModelAndView {  // TODO duplication with RssSubscriptionManager?
            log.info("Got subscription create request: $create")

            val validatedUrl = try {
                URL(create.url)
            } catch (e: MalformedURLException) {
                throw ResponseStatusException(HttpStatus.BAD_REQUEST, "Subscription url is not valid")
            }

            val targetChannel = create.channel
            if (Strings.isNullOrEmpty(targetChannel)) {
                throw ResponseStatusException(HttpStatus.BAD_REQUEST, "Subscription channel missing")
            }
            return conditionalLoads.withChannelForUser(targetChannel, user) { channel ->
                val subscription = RssSubscription(
                    url = validatedUrl.toExternalForm(),
                    channelId = channel.id,
                    username = user.username
                )
                val persisted = subscriptionsDAO.add(subscription)
                log.info("Persisted subscription is: $persisted")
                when (persisted) {
                    is RssSubscription -> rssPoller.requestRead(persisted)
                }
                ModelAndView(viewFactory.jsonView()).addObject("data", persisted)
            }

        }

        return forCurrentUser(::createSubscription)
    }

    @PutMapping("/subscriptions/{id}")
    fun subscriptionUpdate(
        @PathVariable id: String,
        @RequestBody update: SubscriptionUpdateRequest
    ): ModelAndView {
        fun updateSubscription(user: User): ModelAndView {
            return conditionalLoads.withSubscriptionForUser(id, user) { subscription ->
                log.debug("Got subscription update request: {}", update)
                update.name?.let {
                    subscription.name = update.name
                    subscriptionsDAO.save(subscription)
                    log.info("Updated subscription: $subscription")
                }
                ModelAndView(viewFactory.jsonView()).addObject("data", subscription)
            }
        }
        return forCurrentUser(::updateSubscription)
    }

    @GetMapping("/subscriptions/{id}/delete") // TODO should be a HTTP DELETE
    fun deleteSubscription(@PathVariable id: String): ModelAndView {
        fun performDelete(user: User): ModelAndView {
            return conditionalLoads.withSubscriptionForUser(id, user) { subscription ->
                feedItemDAO.deleteSubscriptionFeedItems(subscription)
                subscriptionsDAO.delete(subscription)
                ModelAndView(viewFactory.jsonView()).addObject("data", "ok")
            }
        }
        return forCurrentUser(::performDelete)
    }

}